// @signalapp/libsignal-client/dist/zkgroup/profiles/ClientZkProfileOperations.js
//
// Copyright 2020-2022 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { randomBytes } from 'node:crypto';
import { RANDOM_LENGTH } from '../internal/Constants.js';
import * as Native from '../../Native.js';
import ExpiringProfileKeyCredential from './ExpiringProfileKeyCredential.js';
import ProfileKeyCredentialPresentation from './ProfileKeyCredentialPresentation.js';
import ProfileKeyCredentialRequestContext from './ProfileKeyCredentialRequestContext.js';
export default class ClientZkProfileOperations {
    serverPublicParams;
    constructor(serverPublicParams) {
        this.serverPublicParams = serverPublicParams;
    }
    createProfileKeyCredentialRequestContext(userId, profileKey) {
        const random = randomBytes(RANDOM_LENGTH);
        return this.createProfileKeyCredentialRequestContextWithRandom(random, userId, profileKey);
    }
    createProfileKeyCredentialRequestContextWithRandom(random, userId, profileKey) {
        return new ProfileKeyCredentialRequestContext(Native.ServerPublicParams_CreateProfileKeyCredentialRequestContextDeterministic(this.serverPublicParams, random, userId.getServiceIdFixedWidthBinary(), profileKey.getContents()));
    }
    receiveExpiringProfileKeyCredential(profileKeyCredentialRequestContext, profileKeyCredentialResponse, now = new Date()) {
        return new ExpiringProfileKeyCredential(Native.ServerPublicParams_ReceiveExpiringProfileKeyCredential(this.serverPublicParams, profileKeyCredentialRequestContext.getContents(), profileKeyCredentialResponse.getContents(), Math.floor(now.getTime() / 1000)));
    }
    createExpiringProfileKeyCredentialPresentation(groupSecretParams, profileKeyCredential) {
        const random = randomBytes(RANDOM_LENGTH);
        return this.createExpiringProfileKeyCredentialPresentationWithRandom(random, groupSecretParams, profileKeyCredential);
    }
    createExpiringProfileKeyCredentialPresentationWithRandom(random, groupSecretParams, profileKeyCredential) {
        return new ProfileKeyCredentialPresentation(Native.ServerPublicParams_CreateExpiringProfileKeyCredentialPresentationDeterministic(this.serverPublicParams, random, groupSecretParams.getContents(), profileKeyCredential.getContents()));
    }
}
