// @signalapp/libsignal-client/dist/zkgroup/groups/ClientZkGroupCipher.js
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { randomBytes } from 'node:crypto';
import { RANDOM_LENGTH } from '../internal/Constants.js';
import * as Native from '../../Native.js';
import UuidCiphertext from './UuidCiphertext.js';
import ProfileKeyCiphertext from './ProfileKeyCiphertext.js';
import ProfileKey from '../profiles/ProfileKey.js';
import { ServiceId } from '../../Address.js';
export default class ClientZkGroupCipher {
    groupSecretParams;
    constructor(groupSecretParams) {
        this.groupSecretParams = groupSecretParams;
    }
    encryptServiceId(serviceId) {
        return new UuidCiphertext(Native.GroupSecretParams_EncryptServiceId(this.groupSecretParams.getContents(), serviceId.getServiceIdFixedWidthBinary()));
    }
    decryptServiceId(ciphertext) {
        return ServiceId.parseFromServiceIdFixedWidthBinary(Native.GroupSecretParams_DecryptServiceId(this.groupSecretParams.getContents(), ciphertext.getContents()));
    }
    encryptProfileKey(profileKey, userId) {
        return new ProfileKeyCiphertext(Native.GroupSecretParams_EncryptProfileKey(this.groupSecretParams.getContents(), profileKey.getContents(), userId.getServiceIdFixedWidthBinary()));
    }
    decryptProfileKey(profileKeyCiphertext, userId) {
        return new ProfileKey(Native.GroupSecretParams_DecryptProfileKey(this.groupSecretParams.getContents(), profileKeyCiphertext.getContents(), userId.getServiceIdFixedWidthBinary()));
    }
    encryptBlob(plaintext) {
        const random = randomBytes(RANDOM_LENGTH);
        return this.encryptBlobWithRandom(random, plaintext);
    }
    encryptBlobWithRandom(random, plaintext) {
        return Native.GroupSecretParams_EncryptBlobWithPaddingDeterministic(this.groupSecretParams.getContents(), random, plaintext, 0);
    }
    decryptBlob(blobCiphertext) {
        return Native.GroupSecretParams_DecryptBlobWithPadding(this.groupSecretParams.getContents(), blobCiphertext);
    }
}
