// @signalapp/libsignal-client/dist/zkgroup/groups/GroupSecretParams.js
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { randomBytes } from 'node:crypto';
import ByteArray from '../internal/ByteArray.js';
import * as Native from '../../Native.js';
import { RANDOM_LENGTH } from '../internal/Constants.js';
import GroupMasterKey from './GroupMasterKey.js';
import GroupPublicParams from './GroupPublicParams.js';
export default class GroupSecretParams extends ByteArray {
    __type;
    static generate() {
        const random = randomBytes(RANDOM_LENGTH);
        return GroupSecretParams.generateWithRandom(random);
    }
    static generateWithRandom(random) {
        return new GroupSecretParams(Native.GroupSecretParams_GenerateDeterministic(random));
    }
    static deriveFromMasterKey(groupMasterKey) {
        return new GroupSecretParams(Native.GroupSecretParams_DeriveFromMasterKey(groupMasterKey.getContents()));
    }
    constructor(contents) {
        super(contents, Native.GroupSecretParams_CheckValidContents);
    }
    getMasterKey() {
        return new GroupMasterKey(Native.GroupSecretParams_GetMasterKey(this.contents));
    }
    getPublicParams() {
        return new GroupPublicParams(Native.GroupSecretParams_GetPublicParams(this.contents));
    }
}
