// @signalapp/libsignal-client/dist/net/Chat.js
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import * as Native from '../Native.js';
import * as KT from './KeyTransparency.js';
import { newNativeHandle } from '../internal.js';
import { FakeChatRemote } from './FakeChat.js';
const DEFAULT_CHAT_REQUEST_TIMEOUT_MILLIS = 5000;
export class ChatServerMessageAck {
    _nativeHandle;
    constructor(_nativeHandle) {
        this._nativeHandle = _nativeHandle;
    }
    send(statusCode) {
        Native.ServerMessageAck_SendStatus(this, statusCode);
    }
}
class ConnectionInfoImpl {
    _nativeHandle;
    constructor(_nativeHandle) {
        this._nativeHandle = _nativeHandle;
    }
    get localPort() {
        return Native.ChatConnectionInfo_local_port(this);
    }
    get ipVersion() {
        const value = Native.ChatConnectionInfo_ip_version(this);
        switch (value) {
            case 1:
                return 'IPv4';
            case 2:
                return 'IPv6';
            default:
                throw new TypeError(`ip type was unexpectedly ${value}`);
        }
    }
    toString() {
        return Native.ChatConnectionInfo_description(this);
    }
}
export class UnauthenticatedChatConnection {
    _asyncContext;
    _chatService;
    chatListener;
    env;
    static async connect(asyncContext, connectionManager, listener, env, options) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const connect = Native.UnauthenticatedChatConnection_connect(asyncContext, connectionManager, options?.languages ?? []);
        const chat = await asyncContext.makeCancellable(options?.abortSignal, connect);
        const connection = newNativeHandle(chat);
        Native.UnauthenticatedChatConnection_init_listener(connection, new WeakListenerWrapper(nativeChatListener));
        return new UnauthenticatedChatConnection(asyncContext, connection, nativeChatListener, env);
    }
    /**
     * Creates a chat connection backed by a fake remote end.
     *
     * @param asyncContext the async runtime to use
     * @param listener the listener to send events to
     * @returns an {@link UnauthenticatedChatConnection} and handle for the remote
     * end of the fake connection.
     */
    static fakeConnect(asyncContext, listener) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const fakeChat = newNativeHandle(Native.TESTING_FakeChatConnection_Create(asyncContext, new WeakListenerWrapper(nativeChatListener), ''));
        const chat = newNativeHandle(Native.TESTING_FakeChatConnection_TakeUnauthenticatedChat(fakeChat));
        return [
            new UnauthenticatedChatConnection(asyncContext, chat, nativeChatListener),
            new FakeChatRemote(asyncContext, Native.TESTING_FakeChatConnection_TakeRemote(fakeChat)),
        ];
    }
    constructor(
    // Not true-private so that they can be accessed by the "Service" interfaces in chat/.
    _asyncContext, _chatService, 
    // Unused except to keep the listener alive since the Rust code only holds a
    // weak reference to the same object.
    chatListener, env) {
        this._asyncContext = _asyncContext;
        this._chatService = _chatService;
        this.chatListener = chatListener;
        this.env = env;
    }
    fetch(chatRequest, options) {
        return this._asyncContext.makeCancellable(options?.abortSignal, Native.UnauthenticatedChatConnection_send(this._asyncContext, this._chatService, buildHttpRequest(chatRequest), chatRequest.timeoutMillis ?? DEFAULT_CHAT_REQUEST_TIMEOUT_MILLIS));
    }
    disconnect() {
        return Native.UnauthenticatedChatConnection_disconnect(this._asyncContext, this._chatService);
    }
    connectionInfo() {
        return new ConnectionInfoImpl(Native.UnauthenticatedChatConnection_info(this._chatService));
    }
    keyTransparencyClient() {
        if (this.env == null) {
            throw new Error('KeyTransparency is not supported on local test server');
        }
        return new KT.ClientImpl(this._asyncContext, this._chatService, this.env);
    }
}
export class AuthenticatedChatConnection {
    asyncContext;
    chatService;
    chatListener;
    static async connect(asyncContext, connectionManager, username, password, receiveStories, listener, options) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const connect = Native.AuthenticatedChatConnection_connect(asyncContext, connectionManager, username, password, receiveStories, options?.languages ?? []);
        const chat = await asyncContext.makeCancellable(options?.abortSignal, connect);
        const connection = newNativeHandle(chat);
        Native.AuthenticatedChatConnection_init_listener(connection, new WeakListenerWrapper(nativeChatListener));
        return new AuthenticatedChatConnection(asyncContext, connection, nativeChatListener);
    }
    /**
     * Creates a chat connection backed by a fake remote end.
     *
     * @param asyncContext the async runtime to use
     * @param listener the listener to send events to
     * @param alerts alerts to send immediately upon connect
     * @returns an {@link AuthenticatedChatConnection} and handle for the remote
     * end of the fake connection.
     */
    static fakeConnect(asyncContext, listener, alerts) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const fakeChat = newNativeHandle(Native.TESTING_FakeChatConnection_Create(asyncContext, new WeakListenerWrapper(nativeChatListener), alerts?.join('\n') ?? ''));
        const chat = newNativeHandle(Native.TESTING_FakeChatConnection_TakeAuthenticatedChat(fakeChat));
        return [
            new AuthenticatedChatConnection(asyncContext, chat, nativeChatListener),
            new FakeChatRemote(asyncContext, Native.TESTING_FakeChatConnection_TakeRemote(fakeChat)),
        ];
    }
    constructor(asyncContext, chatService, 
    // Unused except to keep the listener alive since the Rust code only holds a
    // weak reference to the same object.
    chatListener) {
        this.asyncContext = asyncContext;
        this.chatService = chatService;
        this.chatListener = chatListener;
    }
    fetch(chatRequest, options) {
        return this.asyncContext.makeCancellable(options?.abortSignal, Native.AuthenticatedChatConnection_send(this.asyncContext, this.chatService, buildHttpRequest(chatRequest), chatRequest.timeoutMillis ?? DEFAULT_CHAT_REQUEST_TIMEOUT_MILLIS));
    }
    disconnect() {
        return Native.AuthenticatedChatConnection_disconnect(this.asyncContext, this.chatService);
    }
    connectionInfo() {
        return new ConnectionInfoImpl(Native.AuthenticatedChatConnection_info(this.chatService));
    }
}
/**
 * A chat connection used specifically for provisioning linked devices.
 *
 * Note that no messages are sent *from* the client for a provisioning connection; all the
 * interesting functionality is in the events delivered to the {@link ProvisioningConnectionListener}.
 */
export class ProvisioningConnection {
    asyncContext;
    chatService;
    chatListener;
    static async connect(asyncContext, connectionManager, listener, options) {
        const nativeListener = this.makeNativeProvisioningListener(listener);
        const connect = Native.ProvisioningChatConnection_connect(asyncContext, connectionManager);
        const chat = await asyncContext.makeCancellable(options?.abortSignal, connect);
        const connection = newNativeHandle(chat);
        Native.ProvisioningChatConnection_init_listener(connection, new WeakProvisioningListenerWrapper(nativeListener));
        return new ProvisioningConnection(asyncContext, connection, nativeListener);
    }
    /**
     * Creates a provisioning chat connection backed by a fake remote end.
     *
     * @param asyncContext the async runtime to use
     * @param listener the listener to send events to
     * @returns a {@link ProvisioningConnection} and handle for the remote
     * end of the fake connection.
     */
    static fakeConnect(asyncContext, listener) {
        const nativeListener = this.makeNativeProvisioningListener(listener);
        const fakeChat = newNativeHandle(Native.TESTING_FakeChatConnection_CreateProvisioning(asyncContext, new WeakProvisioningListenerWrapper(nativeListener)));
        const chat = newNativeHandle(Native.TESTING_FakeChatConnection_TakeProvisioningChat(fakeChat));
        return [
            new ProvisioningConnection(asyncContext, chat, nativeListener),
            new FakeChatRemote(asyncContext, Native.TESTING_FakeChatConnection_TakeRemote(fakeChat)),
        ];
    }
    static makeNativeProvisioningListener(listener) {
        return {
            receivedAddress(address, ack) {
                listener.onReceivedAddress(address, new ChatServerMessageAck(ack));
            },
            receivedEnvelope(envelope, ack) {
                listener.onReceivedEnvelope(envelope, new ChatServerMessageAck(ack));
            },
            connectionInterrupted(cause) {
                listener.onConnectionInterrupted(cause);
            },
        };
    }
    constructor(asyncContext, chatService, 
    // Unused except to keep the listener alive since the Rust code only holds a
    // weak reference to the same object.
    chatListener) {
        this.asyncContext = asyncContext;
        this.chatService = chatService;
        this.chatListener = chatListener;
    }
    disconnect() {
        return Native.ProvisioningChatConnection_disconnect(this.asyncContext, this.chatService);
    }
    connectionInfo() {
        return new ConnectionInfoImpl(Native.ProvisioningChatConnection_info(this.chatService));
    }
}
/**
 * Holds a {@link Native.ChatListener} by {@link WeakRef} and delegates
 * `ChatListener` calls to it.
 *
 * This lets us avoid passing anything across the bridge that has a normal
 * (strong) reference to the app-side listener. The danger is that the passed-in
 * listener might gain a reference to the JS connection object; that would
 * result in a reference cycle that Node can't clean up because one of the
 * references is through a Rust `Box`.
 *
 * When constructing a connection, calling code should wrap an app-side listener
 * in this type and pass it across the bridge, then hold its own strong
 * reference to the same listener as a field. This ensures that if there is a
 * reference cycle between the connection and app-side listener, that cycle is
 * visible to the Node runtime, while still ensuring the passed-in listener
 * stays alive as long as the connection does.
 */
class WeakListenerWrapper {
    listener;
    constructor(listener) {
        this.listener = new WeakRef(listener);
    }
    connectionInterrupted(reason) {
        this.listener.deref()?.connectionInterrupted(reason);
    }
    receivedIncomingMessage(envelope, timestamp, ack) {
        this.listener.deref()?.receivedIncomingMessage(envelope, timestamp, ack);
    }
    receivedQueueEmpty() {
        this.listener.deref()?.receivedQueueEmpty();
    }
    receivedAlerts(alerts) {
        this.listener.deref()?.receivedAlerts(alerts);
    }
}
/** Like {@link WeakListenerWrapper}, but for {@link ProvisioningConnection}. */
class WeakProvisioningListenerWrapper {
    listener;
    constructor(listener) {
        this.listener = new WeakRef(listener);
    }
    receivedAddress(address, ack) {
        this.listener.deref()?.receivedAddress(address, ack);
    }
    receivedEnvelope(envelope, ack) {
        this.listener.deref()?.receivedEnvelope(envelope, ack);
    }
    connectionInterrupted(reason) {
        this.listener.deref()?.connectionInterrupted(reason);
    }
}
function makeNativeChatListener(asyncContext, listener) {
    if ('onQueueEmpty' in listener) {
        return {
            receivedIncomingMessage(envelope, timestamp, ack) {
                listener.onIncomingMessage(envelope, timestamp, new ChatServerMessageAck(ack));
            },
            receivedQueueEmpty() {
                listener.onQueueEmpty();
            },
            receivedAlerts(alerts) {
                listener.onReceivedAlerts?.(alerts);
            },
            connectionInterrupted(cause) {
                listener.onConnectionInterrupted(cause);
            },
        };
    }
    return {
        receivedIncomingMessage(_envelope, _timestamp, _ack) {
            throw new Error('Event not supported on unauthenticated connection');
        },
        receivedQueueEmpty() {
            throw new Error('Event not supported on unauthenticated connection');
        },
        receivedAlerts(alerts) {
            if (alerts.length != 0) {
                throw new Error(`Got ${alerts.length} unexpected alerts on an unauthenticated connection`);
            }
        },
        connectionInterrupted(cause) {
            listener.onConnectionInterrupted(cause);
        },
    };
}
export function buildHttpRequest(chatRequest) {
    const { verb, path, body, headers } = chatRequest;
    const httpRequest = {
        _nativeHandle: Native.HttpRequest_new(verb, path, body ?? null),
    };
    headers.forEach((header) => {
        const [name, value] = header;
        Native.HttpRequest_add_header(httpRequest, name, value);
    });
    return httpRequest;
}
