// @signalapp/libsignal-client/dist/net/FakeChat.js
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import * as Native from '../Native.js';
import { newNativeHandle } from '../internal.js';
import assert from 'node:assert';
export class InternalRequest {
    _nativeHandle;
    requestId;
    constructor([nativeHandle, requestId]) {
        this._nativeHandle = nativeHandle;
        this.requestId = requestId;
    }
    get verb() {
        return Native.TESTING_ChatRequestGetMethod(this);
    }
    get path() {
        return Native.TESTING_ChatRequestGetPath(this);
    }
    get headers() {
        const names = Native.TESTING_ChatRequestGetHeaderNames(this);
        return new Map(names.map((name) => {
            return [name, Native.TESTING_ChatRequestGetHeaderValue(this, name)];
        }));
    }
    get body() {
        return Native.TESTING_ChatRequestGetBody(this);
    }
}
export class FakeChatRemote {
    asyncContext;
    _nativeHandle;
    static FAKE_AUTH_CONNECT_SELF_UUID = 'ffffffff-ffff-ffff-ffff-ffffffffffff';
    constructor(asyncContext, _nativeHandle) {
        this.asyncContext = asyncContext;
        this._nativeHandle = _nativeHandle;
    }
    async receiveIncomingRequest() {
        const nativeRequest = await Native.TESTING_FakeChatRemoteEnd_ReceiveIncomingRequest(this.asyncContext, this);
        if (nativeRequest === null) {
            return null;
        }
        return new InternalRequest(nativeRequest);
    }
    async assertReceiveIncomingRequest() {
        const request = await this.receiveIncomingRequest();
        assert(request !== null, 'Cannot reply to the request that is null');
        return request;
    }
    sendReplyTo(request, response) {
        const fullResponse = { id: request.requestId, ...response };
        this.sendServerResponse(fullResponse);
    }
    sendServerResponse(response) {
        const message = response.message ?? '';
        const headers = response.headers ?? [];
        const body = response.body ?? null;
        const nativeResponse = newNativeHandle(Native.TESTING_FakeChatResponse_Create(response.id, response.status, message, headers, body));
        Native.TESTING_FakeChatRemoteEnd_SendServerResponse(this, nativeResponse);
    }
    sendRawServerResponse(bytes) {
        Native.TESTING_FakeChatRemoteEnd_SendRawServerResponse(this, bytes);
    }
    sendRawServerRequest(bytes) {
        Native.TESTING_FakeChatRemoteEnd_SendRawServerRequest(this, bytes);
    }
    injectConnectionInterrupted() {
        Native.TESTING_FakeChatRemoteEnd_InjectConnectionInterrupted(this);
    }
}
