// @signalapp/libsignal-client/dist/net/KeyTransparency.js
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import * as Native from '../Native.js';
export class ClientImpl {
    asyncContext;
    chatService;
    env;
    constructor(asyncContext, chatService, env) {
        this.asyncContext = asyncContext;
        this.chatService = chatService;
        this.env = env;
    }
    async check(request, store, options) {
        const { abortSignal } = options ?? {};
        const { aciInfo: { aci, identityKey: aciIdentityKey }, e164Info, usernameHash, mode, } = request;
        const { e164, unidentifiedAccessKey } = e164Info ?? {
            e164: null,
            unidentifiedAccessKey: null,
        };
        const [accountData, newDistinguished] = await this.asyncContext.makeCancellable(abortSignal, Native.KeyTransparency_Check(this.asyncContext, this.env, this.chatService, aci.getServiceIdFixedWidthBinary(), aciIdentityKey, e164, unidentifiedAccessKey, usernameHash ?? null, await store.getAccountData(aci), await store.getLastDistinguishedTreeHead(), mode === 'self', mode === 'self' ? request.isE164Discoverable : true));
        await store.setAccountData(aci, accountData);
        if (newDistinguished.length > 0) {
            await store.setLastDistinguishedTreeHead(newDistinguished);
        }
    }
}
