// @signalapp/libsignal-client/dist/net/Registration.js
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import * as Native from '../Native.js';
import { Aci, Pni, ServiceIdKind } from '../Address.js';
import { newNativeHandle } from '../internal.js';
import { FakeChatRemote } from './FakeChat.js';
/**
 * A client for the Signal registration service.
 *
 * This wraps a {@link Net} to provide a reliable registration service client.
 */
export class RegistrationService {
    _nativeHandle;
    tokioAsyncContext;
    constructor(_nativeHandle, tokioAsyncContext) {
        this._nativeHandle = _nativeHandle;
        this.tokioAsyncContext = tokioAsyncContext;
    }
    /**
     * The stable identifier for the session.
     *
     * This can be persisted and used later for resuming a session that was
     * interrupted.
     */
    get sessionId() {
        return Native.RegistrationService_SessionId(this);
    }
    /**
     * The last known state of the session.
     *
     * The state received from the server is stored internally and is exposed via
     * this property.
     */
    get sessionState() {
        return RegistrationService._convertNativeSessionState(newNativeHandle(Native.RegistrationService_RegistrationSession(this)));
    }
    /**
     * Resumes a previously created registration session.
     *
     * Asynchronously connects to the registration session and verifies that the
     * session is still available. If so, returns an initialized
     * `RegistrationService`. Otherwise the returned `Promise` is resolved with an
     * error.
     *
     * Clients should not use this method directly, but should instead call
     * {@link Net.resumeRegistrationSession}.
     *
     * @returns a `Promise` that resolves to the `RegistrationService` if
     * resumption is successful, otherwise a {@link LibSignalError}.
     */
    static async resumeSession(options, { sessionId, e164 }) {
        const session = await Native.RegistrationService_ResumeSession(options.tokioAsyncContext, sessionId, e164, options.connectionManager);
        return new RegistrationService(session, options.tokioAsyncContext);
    }
    /**
     * Starts a new registration session.
     *
     * Asynchronously connects to the registration session and requests a new session.
     * If successful, returns an initialized `RegistrationService`. Otherwise the
     * returned `Promise` is resolved with an error.
     *
     * Clients should not use this method directly, but should instead call
     * {@link Net.createRegistrationSession}.
     *
     * @returns a `Promise` that resolves to the `RegistrationService` if
     * creation is successful, otherwise a {@link RateLimitedError} or other
     * {@link LibSignalError}.
     */
    static async createSession(options, { e164 }) {
        const session = await Native.RegistrationService_CreateSession(options.tokioAsyncContext, { number: e164 }, options.connectionManager);
        return new RegistrationService(session, options.tokioAsyncContext);
    }
    async submitCaptcha(captcha) {
        await Native.RegistrationService_SubmitCaptcha(this.tokioAsyncContext, this, captcha);
        return this.sessionState;
    }
    /**
     * Request that a verification code be sent via the given transport method.
     *
     * With the websocket transport, this makes a POST request to
     * `/v1/verification/session/{sessionId}/code`.
     *
     * The `languages` parameter should be a list of languages in Accept-Language syntax. Note that
     * "quality weighting" can be left out; the Signal server will always consider the list to be in
     * priority order.
     */
    async requestVerification({ transport, client, languages = [], }) {
        await Native.RegistrationService_RequestVerificationCode(this.tokioAsyncContext, this, transport, client, languages);
    }
    async verifySession(code) {
        await Native.RegistrationService_SubmitVerificationCode(this.tokioAsyncContext, this, code);
        return this.sessionState.verified;
    }
    async checkSvr2Credentials(svr2Tokens) {
        return Native.RegistrationService_CheckSvr2Credentials(this.tokioAsyncContext, this, svr2Tokens);
    }
    async registerAccount(inputs) {
        const request = new RegisterAccountRequest(inputs);
        return new RegisterAccountResponse(await Native.RegistrationService_RegisterAccount(this.tokioAsyncContext, this, request, inputs.accountAttributes));
    }
    static async reregisterAccount(options, inputs) {
        const { tokioAsyncContext, connectionManager } = options;
        const request = new RegisterAccountRequest(inputs);
        return new RegisterAccountResponse(await Native.RegistrationService_ReregisterAccount(tokioAsyncContext, connectionManager, inputs.e164, request, inputs.accountAttributes));
    }
    /**
     *  Internal, only public for testing
     */
    static _convertNativeSessionState(session) {
        const nextCallSecs = Native.RegistrationSession_GetNextCallSeconds(session);
        const nextSmsSecs = Native.RegistrationSession_GetNextSmsSeconds(session);
        const nextVerificationAttemptSecs = Native.RegistrationSession_GetNextVerificationAttemptSeconds(session);
        return {
            allowedToRequestCode: Native.RegistrationSession_GetAllowedToRequestCode(session),
            verified: Native.RegistrationSession_GetVerified(session),
            nextCallSecs: nextCallSecs != null ? nextCallSecs : undefined,
            nextSmsSecs: nextSmsSecs != null ? nextSmsSecs : undefined,
            nextVerificationAttemptSecs: nextVerificationAttemptSecs != null
                ? nextVerificationAttemptSecs
                : undefined,
            requestedInformation: new Set(Native.RegistrationSession_GetRequestedInformation(session)),
        };
    }
    /**
     * Create a registration client that sends requests to the returned fake chat.
     *
     * Calling code will need to await and use the returned fake chat remote
     * to respond in order for the returned Promise<RegistrationService> to resolve.
     *
     * Internal, only public for testing
     */
    static fakeCreateSession(tokio, { e164 }) {
        const server = newNativeHandle(Native.TESTING_FakeChatServer_Create());
        const registration = async () => {
            const handle = await Native.TESTING_FakeRegistrationSession_CreateSession(tokio, { number: e164 }, server);
            return new RegistrationService(handle, tokio);
        };
        const remote = async () => {
            return new FakeChatRemote(tokio, await Native.TESTING_FakeChatServer_GetNextRemote(tokio, server));
        };
        return [registration(), remote()];
    }
}
function toBridgedPublicPreKey(key) {
    return {
        keyId: key.id(),
        signature: key.signature(),
        publicKey: key.publicKey().serialize(),
    };
}
export class AccountAttributes {
    _nativeHandle;
    constructor({ recoveryPassword, aciRegistrationId, pniRegistrationId, registrationLock, unidentifiedAccessKey, unrestrictedUnidentifiedAccess, capabilities, discoverableByPhoneNumber, }) {
        const capabilitiesArray = Array.from(capabilities);
        this._nativeHandle = Native.RegistrationAccountAttributes_Create(recoveryPassword, aciRegistrationId, pniRegistrationId, registrationLock, unidentifiedAccessKey, unrestrictedUnidentifiedAccess, capabilitiesArray, discoverableByPhoneNumber);
    }
}
export class RegisterAccountResponse {
    _nativeHandle;
    constructor(_nativeHandle) {
        this._nativeHandle = _nativeHandle;
    }
    get aci() {
        return new Aci(Native.RegisterAccountResponse_GetIdentity(this, ServiceIdKind.Aci));
    }
    get pni() {
        return new Pni(Native.RegisterAccountResponse_GetIdentity(this, ServiceIdKind.Pni));
    }
    get number() {
        return Native.RegisterAccountResponse_GetNumber(this);
    }
    get usernameHash() {
        return Native.RegisterAccountResponse_GetUsernameHash(this);
    }
    get usernameLinkHandle() {
        return Native.RegisterAccountResponse_GetUsernameLinkHandle(this);
    }
    get backupEntitlement() {
        const backupLevel = Native.RegisterAccountResponse_GetEntitlementBackupLevel(this);
        const expirationSeconds = Native.RegisterAccountResponse_GetEntitlementBackupExpirationSeconds(this);
        if (backupLevel == null || expirationSeconds == null)
            return null;
        return {
            backupLevel,
            expirationSeconds,
        };
    }
    get entitlementBadges() {
        return Native.RegisterAccountResponse_GetEntitlementBadges(this);
    }
    get reregistration() {
        return Native.RegisterAccountResponse_GetReregistration(this);
    }
    get storageCapable() {
        return Native.RegisterAccountResponse_GetStorageCapable(this);
    }
}
class RegisterAccountRequest {
    _nativeHandle;
    constructor(inputs) {
        const { accountPassword, skipDeviceTransfer, aciPublicKey, pniPublicKey, aciSignedPreKey, pniSignedPreKey, aciPqLastResortPreKey, pniPqLastResortPreKey, } = inputs;
        this._nativeHandle = Native.RegisterAccountRequest_Create();
        Native.RegisterAccountRequest_SetAccountPassword(this, accountPassword);
        if (skipDeviceTransfer) {
            Native.RegisterAccountRequest_SetSkipDeviceTransfer(this);
        }
        Native.RegisterAccountRequest_SetIdentityPublicKey(this, ServiceIdKind.Aci, aciPublicKey);
        Native.RegisterAccountRequest_SetIdentityPublicKey(this, ServiceIdKind.Pni, pniPublicKey);
        Native.RegisterAccountRequest_SetIdentitySignedPreKey(this, ServiceIdKind.Aci, toBridgedPublicPreKey(aciSignedPreKey));
        Native.RegisterAccountRequest_SetIdentitySignedPreKey(this, ServiceIdKind.Pni, toBridgedPublicPreKey(pniSignedPreKey));
        Native.RegisterAccountRequest_SetIdentityPqLastResortPreKey(this, ServiceIdKind.Aci, toBridgedPublicPreKey(aciPqLastResortPreKey));
        Native.RegisterAccountRequest_SetIdentityPqLastResortPreKey(this, ServiceIdKind.Pni, toBridgedPublicPreKey(pniPqLastResortPreKey));
    }
}
