// @signalapp/libsignal-client/dist/zkgroup/receipts/ClientZkReceiptOperations.js
//
// Copyright 2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { randomBytes } from 'node:crypto';
import { RANDOM_LENGTH } from '../internal/Constants.js';
import * as Native from '../../Native.js';
import ReceiptCredential from './ReceiptCredential.js';
import ReceiptCredentialPresentation from './ReceiptCredentialPresentation.js';
import ReceiptCredentialRequestContext from './ReceiptCredentialRequestContext.js';
export default class ClientZkReceiptOperations {
    serverPublicParams;
    constructor(serverPublicParams) {
        this.serverPublicParams = serverPublicParams;
    }
    createReceiptCredentialRequestContext(receiptSerial) {
        const random = randomBytes(RANDOM_LENGTH);
        return this.createReceiptCredentialRequestContextWithRandom(random, receiptSerial);
    }
    createReceiptCredentialRequestContextWithRandom(random, receiptSerial) {
        return new ReceiptCredentialRequestContext(Native.ServerPublicParams_CreateReceiptCredentialRequestContextDeterministic(this.serverPublicParams, random, receiptSerial.getContents()));
    }
    receiveReceiptCredential(receiptCredentialRequestContext, receiptCredentialResponse) {
        return new ReceiptCredential(Native.ServerPublicParams_ReceiveReceiptCredential(this.serverPublicParams, receiptCredentialRequestContext.getContents(), receiptCredentialResponse.getContents()));
    }
    createReceiptCredentialPresentation(receiptCredential) {
        const random = randomBytes(RANDOM_LENGTH);
        return this.createReceiptCredentialPresentationWithRandom(random, receiptCredential);
    }
    createReceiptCredentialPresentationWithRandom(random, receiptCredential) {
        return new ReceiptCredentialPresentation(Native.ServerPublicParams_CreateReceiptCredentialPresentationDeterministic(this.serverPublicParams, random, receiptCredential.getContents()));
    }
}
