// @signalapp/libsignal-client/dist/zkgroup/backups/BackupAuthCredential.js
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { randomBytes } from 'node:crypto';
import * as Native from '../../Native.js';
import ByteArray from '../internal/ByteArray.js';
import { RANDOM_LENGTH } from '../internal/Constants.js';
import BackupAuthCredentialPresentation from './BackupAuthCredentialPresentation.js';
import BackupLevel from './BackupLevel.js';
import BackupCredentialType from './BackupCredentialType.js';
export default class BackupAuthCredential extends ByteArray {
    __type;
    constructor(contents) {
        super(contents, Native.BackupAuthCredential_CheckValidContents);
    }
    present(serverParams) {
        const random = randomBytes(RANDOM_LENGTH);
        return this.presentWithRandom(serverParams, random);
    }
    presentWithRandom(serverParams, random) {
        return new BackupAuthCredentialPresentation(Native.BackupAuthCredential_PresentDeterministic(this.contents, serverParams.contents, random));
    }
    getBackupId() {
        return Native.BackupAuthCredential_GetBackupId(this.contents);
    }
    getBackupLevel() {
        const n = Native.BackupAuthCredential_GetBackupLevel(this.contents);
        if (!(n in BackupLevel)) {
            throw new TypeError(`Invalid BackupLevel ${n}`);
        }
        return n;
    }
    getType() {
        const n = Native.BackupAuthCredential_GetType(this.contents);
        if (!(n in BackupCredentialType)) {
            throw new TypeError(`Invalid BackupCredentialType ${n}`);
        }
        return n;
    }
}
