// @signalapp/libsignal-client/dist/zkgroup/ServerSecretParams.js
//
// Copyright 2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { randomBytes } from 'node:crypto';
import * as Native from '../Native.js';
import { RANDOM_LENGTH } from './internal/Constants.js';
import ServerPublicParams from './ServerPublicParams.js';
import NotarySignature from './NotarySignature.js';
export default class ServerSecretParams {
    static generate() {
        const random = randomBytes(RANDOM_LENGTH);
        return ServerSecretParams.generateWithRandom(random);
    }
    static generateWithRandom(random) {
        return new ServerSecretParams(Native.ServerSecretParams_GenerateDeterministic(random));
    }
    _nativeHandle;
    constructor(contents) {
        if (contents instanceof Uint8Array) {
            this._nativeHandle = Native.ServerSecretParams_Deserialize(contents);
        }
        else {
            this._nativeHandle = contents;
        }
    }
    getPublicParams() {
        return new ServerPublicParams(Native.ServerSecretParams_GetPublicParams(this));
    }
    sign(message) {
        const random = randomBytes(RANDOM_LENGTH);
        return this.signWithRandom(random, message);
    }
    signWithRandom(random, message) {
        return new NotarySignature(Native.ServerSecretParams_SignDeterministic(this, random, message));
    }
    serialize() {
        return Native.ServerSecretParams_Serialize(this);
    }
}
