// @signalapp/libsignal-client/dist/Errors.js
//
// Copyright 2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { ProtocolAddress, ServiceId } from './Address.js';
export var ErrorCode;
(function (ErrorCode) {
    ErrorCode[ErrorCode["Generic"] = 0] = "Generic";
    ErrorCode[ErrorCode["DuplicatedMessage"] = 1] = "DuplicatedMessage";
    ErrorCode[ErrorCode["SealedSenderSelfSend"] = 2] = "SealedSenderSelfSend";
    ErrorCode[ErrorCode["UntrustedIdentity"] = 3] = "UntrustedIdentity";
    ErrorCode[ErrorCode["InvalidRegistrationId"] = 4] = "InvalidRegistrationId";
    ErrorCode[ErrorCode["InvalidProtocolAddress"] = 5] = "InvalidProtocolAddress";
    ErrorCode[ErrorCode["VerificationFailed"] = 6] = "VerificationFailed";
    ErrorCode[ErrorCode["InvalidSession"] = 7] = "InvalidSession";
    ErrorCode[ErrorCode["InvalidSenderKeySession"] = 8] = "InvalidSenderKeySession";
    ErrorCode[ErrorCode["NicknameCannotBeEmpty"] = 9] = "NicknameCannotBeEmpty";
    ErrorCode[ErrorCode["CannotStartWithDigit"] = 10] = "CannotStartWithDigit";
    ErrorCode[ErrorCode["MissingSeparator"] = 11] = "MissingSeparator";
    ErrorCode[ErrorCode["BadNicknameCharacter"] = 12] = "BadNicknameCharacter";
    ErrorCode[ErrorCode["NicknameTooShort"] = 13] = "NicknameTooShort";
    ErrorCode[ErrorCode["NicknameTooLong"] = 14] = "NicknameTooLong";
    ErrorCode[ErrorCode["DiscriminatorCannotBeEmpty"] = 15] = "DiscriminatorCannotBeEmpty";
    ErrorCode[ErrorCode["DiscriminatorCannotBeZero"] = 16] = "DiscriminatorCannotBeZero";
    ErrorCode[ErrorCode["DiscriminatorCannotBeSingleDigit"] = 17] = "DiscriminatorCannotBeSingleDigit";
    ErrorCode[ErrorCode["DiscriminatorCannotHaveLeadingZeros"] = 18] = "DiscriminatorCannotHaveLeadingZeros";
    ErrorCode[ErrorCode["BadDiscriminatorCharacter"] = 19] = "BadDiscriminatorCharacter";
    ErrorCode[ErrorCode["DiscriminatorTooLarge"] = 20] = "DiscriminatorTooLarge";
    ErrorCode[ErrorCode["IoError"] = 21] = "IoError";
    ErrorCode[ErrorCode["CdsiInvalidToken"] = 22] = "CdsiInvalidToken";
    ErrorCode[ErrorCode["InvalidUri"] = 23] = "InvalidUri";
    ErrorCode[ErrorCode["InvalidMediaInput"] = 24] = "InvalidMediaInput";
    ErrorCode[ErrorCode["UnsupportedMediaInput"] = 25] = "UnsupportedMediaInput";
    ErrorCode[ErrorCode["InputDataTooLong"] = 26] = "InputDataTooLong";
    ErrorCode[ErrorCode["InvalidEntropyDataLength"] = 27] = "InvalidEntropyDataLength";
    ErrorCode[ErrorCode["InvalidUsernameLinkEncryptedData"] = 28] = "InvalidUsernameLinkEncryptedData";
    ErrorCode[ErrorCode["RateLimitedError"] = 29] = "RateLimitedError";
    ErrorCode[ErrorCode["RateLimitChallengeError"] = 30] = "RateLimitChallengeError";
    ErrorCode[ErrorCode["SvrDataMissing"] = 31] = "SvrDataMissing";
    ErrorCode[ErrorCode["SvrRequestFailed"] = 32] = "SvrRequestFailed";
    ErrorCode[ErrorCode["SvrRestoreFailed"] = 33] = "SvrRestoreFailed";
    ErrorCode[ErrorCode["SvrAttestationError"] = 34] = "SvrAttestationError";
    ErrorCode[ErrorCode["SvrInvalidData"] = 35] = "SvrInvalidData";
    ErrorCode[ErrorCode["ChatServiceInactive"] = 36] = "ChatServiceInactive";
    ErrorCode[ErrorCode["AppExpired"] = 37] = "AppExpired";
    ErrorCode[ErrorCode["DeviceDelinked"] = 38] = "DeviceDelinked";
    ErrorCode[ErrorCode["ConnectionInvalidated"] = 39] = "ConnectionInvalidated";
    ErrorCode[ErrorCode["ConnectedElsewhere"] = 40] = "ConnectedElsewhere";
    ErrorCode[ErrorCode["PossibleCaptiveNetwork"] = 41] = "PossibleCaptiveNetwork";
    ErrorCode[ErrorCode["BackupValidation"] = 42] = "BackupValidation";
    ErrorCode[ErrorCode["Cancelled"] = 43] = "Cancelled";
    ErrorCode[ErrorCode["KeyTransparencyError"] = 44] = "KeyTransparencyError";
    ErrorCode[ErrorCode["KeyTransparencyVerificationFailed"] = 45] = "KeyTransparencyVerificationFailed";
    ErrorCode[ErrorCode["IncrementalMacVerificationFailed"] = 46] = "IncrementalMacVerificationFailed";
    ErrorCode[ErrorCode["RequestUnauthorized"] = 47] = "RequestUnauthorized";
    ErrorCode[ErrorCode["MismatchedDevices"] = 48] = "MismatchedDevices";
    ErrorCode[ErrorCode["ServiceIdNotFound"] = 49] = "ServiceIdNotFound";
    ErrorCode[ErrorCode["UploadTooLarge"] = 50] = "UploadTooLarge";
})(ErrorCode || (ErrorCode = {}));
/**
 * A failure sending to a recipient on account of not being up to date on their devices.
 *
 * An entry in {@link MismatchedDevicesError}. Each entry represents a recipient that has either
 * added, removed, or relinked some devices in their account (potentially including their primary
 * device), as represented by the {@link MismatchedDevicesEntry#missingDevices},
 * {@link MismatchedDevicesEntry#extraDevices}, and {@link MismatchedDevicesEntry#staleDevices}
 * arrays, respectively. Handling the exception involves removing the "extra" devices and
 * establishing new sessions for the "missing" and "stale" devices.
 */
export class MismatchedDevicesEntry {
    account;
    missingDevices;
    extraDevices;
    staleDevices;
    constructor({ account, missingDevices, extraDevices, staleDevices, }) {
        this.account =
            account instanceof ServiceId
                ? account
                : ServiceId.parseFromServiceIdFixedWidthBinary(account);
        this.missingDevices = missingDevices ?? [];
        this.extraDevices = extraDevices ?? [];
        this.staleDevices = staleDevices ?? [];
    }
}
export class LibSignalErrorBase extends Error {
    code;
    operation;
    _addr;
    constructor(message, name, operation, extraProps) {
        super(message);
        // Include the dynamic check for `name in ErrorCode` in case there's a bug in the Rust code.
        if (name !== undefined && name in ErrorCode) {
            this.name = name;
            this.code = ErrorCode[name];
        }
        else {
            this.name = 'LibSignalError';
            this.code = ErrorCode.Generic;
        }
        this.operation = operation;
        if (extraProps !== undefined) {
            Object.assign(this, extraProps);
        }
        // Maintains proper stack trace, where our error was thrown (only available on V8)
        //   via https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this);
        }
    }
    get addr() {
        switch (this.code) {
            case ErrorCode.UntrustedIdentity:
                return this._addr;
            case ErrorCode.InvalidRegistrationId:
                return ProtocolAddress._fromNativeHandle(this._addr);
            default:
                throw new TypeError(`cannot get address from this error (${this})`);
        }
    }
    toString() {
        return `${this.name} - ${this.operation}: ${this.message}`;
    }
    /// Like `error.code === code`, but also providing access to any additional properties.
    is(code) {
        return this.code === code;
    }
    /// Like `error instanceof LibSignalErrorBase && error.code === code`, but all in one expression,
    /// and providing access to any additional properties.
    static is(error, code) {
        if (error instanceof LibSignalErrorBase) {
            return error.is(code);
        }
        return false;
    }
}
