// @signalapp/libsignal-client/dist/incremental_mac.js
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import * as stream from 'node:stream';
import { Buffer } from 'node:buffer';
import * as Native from './Native.js';
import { LibSignalErrorBase, } from './Errors.js';
export function everyNthByte(n) {
    return { kind: 'everyN', n: n };
}
export function inferChunkSize(dataSize) {
    return { kind: 'chunksOf', dataSize: dataSize };
}
class DigestingWritable extends stream.Writable {
    _nativeHandle;
    _digests = [];
    constructor(key, sizeChoice) {
        super();
        this._nativeHandle = Native.IncrementalMac_Initialize(key, chunkSizeInBytes(sizeChoice));
    }
    getFinalDigest() {
        // Use Buffer.concat for convenience, but return a proper Uint8Array, both for the correct type
        // and to make an independent copy of a possibly-reused buffer.
        return new Uint8Array(Buffer.concat(this._digests));
    }
    _write(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    chunk, encoding, callback) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        const buffer = Buffer.from(chunk, encoding);
        const next_digest = Native.IncrementalMac_Update(this, buffer, 0, buffer.length);
        if (next_digest.length != 0) {
            this._digests.push(next_digest);
        }
        callback();
    }
    _final(callback) {
        this._digests.push(Native.IncrementalMac_Finalize(this));
        callback();
    }
}
export class DigestingPassThrough extends stream.Transform {
    digester;
    constructor(key, sizeChoice) {
        super();
        this.digester = new DigestingWritable(key, sizeChoice);
        // We handle errors coming from write/end
        this.digester.on('error', () => {
            /* noop */
        });
    }
    getFinalDigest() {
        return this.digester.getFinalDigest();
    }
    _transform(data, enc, callback) {
        this.digester.write(data, enc, (err) => {
            if (err) {
                return callback(err);
            }
            this.push(data);
            callback();
        });
    }
    _final(callback) {
        this.digester.end((err) => {
            if (err) {
                return callback(err);
            }
            callback();
        });
    }
}
class ValidatingWritable extends stream.Writable {
    _nativeHandle;
    _validatedBytes = 0;
    constructor(key, sizeChoice, digest) {
        super();
        const handle = Native.ValidatingMac_Initialize(key, chunkSizeInBytes(sizeChoice), digest);
        if (!handle) {
            // Not sure why eslint isn't treating IncrementalMacVerificationFailed as an Error;
            // standalone examples are not reproducing.
            // eslint-disable-next-line @typescript-eslint/only-throw-error
            throw makeVerificationError('Invalid configuration data');
        }
        this._nativeHandle = handle;
    }
    validatedSize() {
        return this._validatedBytes;
    }
    _write(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    chunk, encoding, callback) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        const buffer = Buffer.from(chunk, encoding);
        const validBytes = Native.ValidatingMac_Update(this, buffer, 0, buffer.length);
        if (validBytes >= 0) {
            this._validatedBytes += validBytes;
            callback();
        }
        else {
            callback(makeVerificationError('Corrupted input data'));
        }
    }
    _final(callback) {
        const validBytes = Native.ValidatingMac_Finalize(this);
        if (validBytes >= 0) {
            this._validatedBytes += validBytes;
            callback();
        }
        else {
            callback(makeVerificationError('Corrupted input data (finalize)'));
        }
    }
}
export class ValidatingPassThrough extends stream.Transform {
    validator;
    buffer = new Array();
    constructor(key, sizeChoice, digest) {
        super();
        this.validator = new ValidatingWritable(key, sizeChoice, digest);
        // We handle errors coming from write/end
        this.validator.on('error', () => {
            /* noop */
        });
    }
    _transform(data, enc, callback) {
        const start = this.validator.validatedSize();
        this.validator.write(data, enc, (err) => {
            if (err) {
                return callback(err);
            }
            this.buffer.push(data);
            const end = this.validator.validatedSize();
            const readySize = end - start;
            // Fully buffer
            if (readySize === 0) {
                return callback(null);
            }
            const { buffer } = this;
            this.buffer = [];
            let validated = 0;
            for (const chunk of buffer) {
                validated += chunk.byteLength;
                // Buffered chunk is fully validated - push it without slicing
                if (validated <= readySize) {
                    this.push(chunk);
                    continue;
                }
                // Validation boundary lies within the chunk, split it
                const notValidated = validated - readySize;
                this.push(chunk.subarray(0, -notValidated));
                this.buffer.push(chunk.subarray(-notValidated));
                // Technically this chunk must be the last chunk so we could break,
                // but for consistency keep looping.
            }
            callback(null);
        });
    }
    _final(callback) {
        const start = this.validator.validatedSize();
        this.validator.end((err) => {
            if (err) {
                return callback(err);
            }
            const end = this.validator.validatedSize();
            const readySize = end - start;
            const buffer = Buffer.concat(this.buffer);
            this.buffer = [];
            if (buffer.byteLength !== readySize) {
                return callback(new Error('Stream not fully processed'));
            }
            this.push(buffer);
            callback(null);
        });
    }
}
export function chunkSizeInBytes(sizeChoice) {
    switch (sizeChoice.kind) {
        case 'everyN':
            return sizeChoice.n;
        case 'chunksOf':
            return Native.IncrementalMac_CalculateChunkSize(sizeChoice.dataSize);
    }
}
function makeVerificationError(message) {
    return new LibSignalErrorBase(message, 'IncrementalMacVerificationFailed', 'incremental_mac');
}
