// @signalapp/libsignal-client/dist/usernames.js
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
/* eslint @typescript-eslint/no-shadow: ["error", { "allow": ["hash"] }] */
import { randomBytes } from 'node:crypto';
import { RANDOM_LENGTH } from './zkgroup/internal/Constants.js';
import * as Native from './Native.js';
export function generateCandidates(nickname, minNicknameLength, maxNicknameLength) {
    return Native.Username_CandidatesFrom(nickname, minNicknameLength, maxNicknameLength);
}
export function fromParts(nickname, discriminator, minNicknameLength, maxNicknameLength) {
    const hash = Native.Username_HashFromParts(nickname, discriminator, minNicknameLength, maxNicknameLength);
    // If we generated the hash correctly, we can format the nickname and discriminator manually.
    const username = `${nickname}.${discriminator}`;
    return { username, hash };
}
export function hash(username) {
    return Native.Username_Hash(username);
}
export function generateProof(username) {
    const random = randomBytes(RANDOM_LENGTH);
    return generateProofWithRandom(username, random);
}
export function generateProofWithRandom(username, random) {
    return Native.Username_Proof(username, random);
}
export function decryptUsernameLink(usernameLink) {
    return Native.UsernameLink_DecryptUsername(usernameLink.entropy, usernameLink.encryptedUsername);
}
export function createUsernameLink(username, previousEntropy) {
    const usernameLinkData = Native.UsernameLink_Create(username, previousEntropy ?? null);
    const entropy = usernameLinkData.subarray(0, 32);
    const encryptedUsername = usernameLinkData.subarray(32);
    return { entropy, encryptedUsername };
}
// Only for testing. Will throw on failure.
export function verifyProof(proof, hash) {
    Native.Username_Verify(proof, hash);
}
